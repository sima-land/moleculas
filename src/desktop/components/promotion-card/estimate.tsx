import React from 'react';
import { formatISO, getMonth, differenceInDays, getDate, getYear } from 'date-fns';
import { Timer } from '@dev-dep/ui-nucleons/timer';

export interface EstimateProps {
  dueDate: Date
}

const MONTH_NAME = [
  'янв',
  'фев',
  'мар',
  'апр',
  'мая',
  'июн',
  'июл',
  'авг',
  'сен',
  'окт',
  'ноя',
  'дек',
] as const;

const toTimePart = (n: number): string => `${n}`.padStart(2, '0');

export const Estimate = ({ dueDate }: EstimateProps) => {
  const now = new Date();

  let content: React.ReactNode = null;

  if (differenceInDays(dueDate, now) < 1) {
    content = (
      <Timer
        date={formatISO(dueDate)}
        format={
          d => [
            d.hours % 24,
            d.minutes % 60,
            d.seconds % 60,
          ].map(toTimePart).join(' : ')
        }
      />
    );
  } else if (getYear(dueDate) !== getYear(now)) {
    const text = [getDate, getMonth, getYear]
      .map(fn => toTimePart(fn(dueDate)))
      .join('.');

    content = `До ${text}`;
  } else {
    content = `До ${getMonth(dueDate)} ${MONTH_NAME[getMonth(dueDate)]}`;
  }

  return (
    <>{content}</>
  );
};
