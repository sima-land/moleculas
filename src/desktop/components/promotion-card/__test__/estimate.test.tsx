import React from 'react';
import { mount } from 'enzyme';
import { Estimate, EstimateProps } from '../estimate';
import { addDays, addHours, addYears } from 'date-fns';

describe('<Estimate />', () => {
  beforeEach(() => {
    jest.useFakeTimers('modern');
    jest.setSystemTime(new Date('2021-01-01T09:00:00'));
  });

  it('should renders correctly', () => {
    const variants: EstimateProps[] = [
      {
        dueDate: addHours(new Date(), 3),
      },
      {
        dueDate: addYears(new Date(), 5),
      },
      {
        dueDate: addDays(new Date(), 7),
      },
    ];

    variants.forEach(variant => {
      const wrapper = mount(
        <Estimate {...variant} />
      );

      expect(wrapper).toMatchSnapshot();
    });
  });
});
