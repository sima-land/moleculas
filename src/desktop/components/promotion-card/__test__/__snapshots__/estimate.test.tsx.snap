// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Estimate /> should renders correctly 1`] = `
<Estimate
  dueDate={2021-01-01T12:00:00.000Z}
>
  <Timer
    date="2021-01-01T12:00:00Z"
    format={[Function]}
  >
    03 : 00 : 00
  </Timer>
</Estimate>
`;

exports[`<Estimate /> should renders correctly 2`] = `
<Estimate
  dueDate={2026-01-01T09:00:00.000Z}
>
  До 01.00.2026
</Estimate>
`;

exports[`<Estimate /> should renders correctly 3`] = `
<Estimate
  dueDate={2021-01-08T09:00:00.000Z}
>
  До 0 янв
</Estimate>
`;
